File: examples/ping_distributed/cpp/ping_distributed.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <algorithm>
#include <iostream>
#include <string>
#include <utility>
#include <vector>

#include <holoscan/holoscan.hpp>
#include <holoscan/operators/ping_tensor_rx/ping_tensor_rx.hpp>
#include <holoscan/operators/ping_tensor_tx/ping_tensor_tx.hpp>

class Fragment1 : public holoscan::Fragment {
 public:
  // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
  Fragment1(bool gpu_tensor, int64_t count, int32_t batch_size, int32_t rows, int32_t columns,
            int32_t channels, std::string data_type)
      : gpu_tensor_(gpu_tensor),
        batch_size_(batch_size),
        count_(count),
        rows_(rows),
        columns_(columns),
        channels_(channels),
        data_type_(std::move(data_type)) {}

  void compose() override {
    using namespace holoscan;

    auto tx = make_operator<ops::PingTensorTxOp>(
        "tx",
        make_condition<CountCondition>(count_),
        Arg("storage_type", std::string{gpu_tensor_ ? "device" : "system"}),
        Arg("batch_size", batch_size_),
        Arg("rows", rows_),
        Arg("columns", columns_),
        Arg("channels", channels_),
        Arg("data_type", data_type_));
    add_operator(tx);
  }

 private:
  bool gpu_tensor_ = false;
  int64_t count_ = 10;
  int32_t batch_size_ = 0;
  int32_t rows_ = 32;
  int32_t columns_ = 64;
  int32_t channels_ = 0;
  std::string data_type_{"uint8_t"};
};

class Fragment2 : public holoscan::Fragment {
 public:
  Fragment2() = default;

  void compose() override {
    using namespace holoscan;
    auto rx = make_operator<ops::PingTensorRxOp>("rx");
    add_operator(rx);
  }
};

class App : public holoscan::Application {
 public:
  // Inherit the constructor
  using Application::Application;

  // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
  void set_options(bool gpu_tensor = false, int64_t count = 10, int32_t batch_size = 0,
                   int32_t rows = 32, int32_t columns = 1024, int32_t channels = 0,
                   const std::string& data_type = "uint8_t") {
    HOLOSCAN_LOG_INFO("Configuring application to use {} tensors", gpu_tensor ? "GPU" : "host");
    gpu_tensor_ = gpu_tensor;
    count_ = count;
    batch_size_ = batch_size;
    rows_ = rows;
    columns_ = columns;
    channels_ = channels;
    data_type_ = data_type;
  }

  void compose() override {
    using namespace holoscan;
    auto fragment1 = make_fragment<Fragment1>(
        "fragment1", gpu_tensor_, count_, batch_size_, rows_, columns_, channels_, data_type_);
    auto fragment2 = make_fragment<Fragment2>("fragment2");

    // Connect the two fragments (tx.out -> rx.in)
    // We can skip the "out" and "in" suffixes, as they are the default
    add_flow(fragment1, fragment2, {{"tx", "rx"}});
  }

 private:
  bool gpu_tensor_ = false;
  int64_t count_ = 10;
  int32_t batch_size_ = 0;
  int32_t rows_ = 32;
  int32_t columns_ = 64;
  int32_t channels_ = 0;
  std::string data_type_{"uint8_t"};
};

std::optional<bool> get_boolean_arg(std::vector<std::string> args, const std::string& name) {
  if (std::find(args.begin(), args.end(), name) != std::end(args)) {
    return true;
  }
  return {};
}

std::optional<int32_t> get_int32_arg(std::vector<std::string> args, const std::string& name) {
  auto loc = std::find(args.begin(), args.end(), name);
  if ((loc != std::end(args)) && (++loc != std::end(args))) {
    try {
      return std::stoi(*loc);
    } catch (std::exception& e) {
      HOLOSCAN_LOG_ERROR("Unable to parse provided argument '{}'", name);
      return {};
    }
  }
  return {};
}
std::optional<int64_t> get_int64_arg(std::vector<std::string> args, const std::string& name) {
  auto loc = std::find(args.begin(), args.end(), name);
  if ((loc != std::end(args)) && (++loc != std::end(args))) {
    try {
      return std::stoll(*loc);
    } catch (std::exception& e) {
      HOLOSCAN_LOG_ERROR("Unable to parse provided argument '{}'", name);
      return {};
    }
  }
  return {};
}

std::optional<std::string> get_str_arg(std::vector<std::string> args, const std::string& name) {
  auto loc = std::find(args.begin(), args.end(), name);
  if ((loc != std::end(args)) && (++loc != std::end(args))) {
    return *loc;
  }
  return {};
}

int main() {
  // Print info on the additional command line options supported by this app.
  // Note: -h and --help are intercepted by the Application class to handle the built-in options
  //       for distributed apps.
  std::cout << "Additional supported arguments: \n"
            << "  --gpu               If specified, tensors will be in GPU memory.\n"
            << "  --count COUNT       The number of times to send the tensor.\n"
            << "  --batch_size BATCH  The batch size of the tensor (dimension omitted if 0).\n"
            << "  --rows ROWS         The number of rows in the tensor.\n"
            << "  --columns COLUMNS   The number of columns (dimension omitted if 0).\n"
            << "  --channels CHANNELS The number of channels (dimension omitted if 0).\n"
            << "  --data_type TYPE    The C++ type of the data elements. Must be one of \n"
            << "                      {'int8_t', 'int16_t', 'int32_t', 'int64_t',  \n"
            << "                       'uint8_t', 'uint16_t', 'uint32_t', 'uint64_t',  \n"
            << "                       'float', 'double', 'complex<float>', complex<double>'}.\n"
            << "  --track             If specified, data flow tracking will be enabled.\n"
            << std::endl;

  auto app = holoscan::make_application<App>();

  // Parse args that are defined for all applications.
  std::vector<std::string>& remaining_args = app->argv();

  // Parse any additional supported arguments
  bool tensor_on_gpu = get_boolean_arg(remaining_args, "--gpu").value_or(false);
  int64_t count = get_int64_arg(remaining_args, "--count").value_or(10);
  int32_t batch_size = get_int32_arg(remaining_args, "--batch_size").value_or(0);
  int32_t rows = get_int32_arg(remaining_args, "--rows").value_or(32);
  int32_t columns = get_int32_arg(remaining_args, "--columns").value_or(64);
  int32_t channels = get_int32_arg(remaining_args, "--channels").value_or(0);
  std::string data_type = get_str_arg(remaining_args, "--data_type").value_or("uint8_t");
  bool data_flow_tracking_enabled = get_boolean_arg(remaining_args, "--track").value_or(false);

  HOLOSCAN_LOG_INFO("Running ping with tensors on {}.", tensor_on_gpu ? "GPU" : "host");

  // configure tensor on host vs. GPU and set the count and shape
  app->set_options(tensor_on_gpu, count, batch_size, rows, columns, channels, data_type);

  if (data_flow_tracking_enabled) {
    // enable data flow tracking for a distributed app
    auto trackers = app->track_distributed(0, 0, 0);

    // run the application
    app->run();

    // print data flow tracking results
    for (const auto& [name, tracker] : trackers) {
      std::cout << "Fragment: " << name << std::endl;
      tracker->print();
    }
  } else {
    app->run();
  }

  return 0;
}

File: examples/ping_distributed/python/ping_distributed.py
"""
SPDX-FileCopyrightText: Copyright (c) 2023-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""  # noqa: E501

import textwrap
from argparse import ArgumentParser

import numpy as np

from holoscan.conditions import CountCondition
from holoscan.core import Application, Fragment, Tracker
from holoscan.operators import PingTensorRxOp, PingTensorTxOp


class Fragment1(Fragment):
    def __init__(
        self,
        *args,
        gpu_tensor=False,
        count=10,
        shape=(32, 64),
        dtype="uint8_t",
        tensor_name="out",
        **kwargs,
    ):
        self.gpu_tensor = gpu_tensor
        self.count = count
        if len(shape) == 2:
            self.rows, self.columns = shape
        else:
            raise ValueError("expected shape of length 2")
        self.shape = shape
        self.dtype = dtype
        self.tensor_name = tensor_name
        super().__init__(*args, **kwargs)

    def compose(self):
        # Configure the operators. Here we use CountCondition to terminate
        # execution after a specific number of messages have been sent.
        storage_type = "device" if self.gpu_tensor else "system"
        tx = PingTensorTxOp(
            self,
            CountCondition(self, self.count),
            storage_type=storage_type,
            rows=self.rows,
            columns=self.columns,
            dtype=self.dtype,
            tensor_name=self.tensor_name,
            name="tx",
        )

        # Add the operator (tx) to the fragment
        self.add_operator(tx)


class Fragment2(Fragment):
    def compose(self):
        rx = PingTensorRxOp(self, name="rx")

        # Add the operator (rx) to the fragment
        self.add_operator(rx)


class MyPingApp(Application):
    def __init__(self, *args, gpu_tensor=False, count=10, shape=(32, 64), dtype=np.uint8, **kwargs):
        self.gpu_tensor = gpu_tensor
        self.count = count
        self.shape = shape
        self.dtype = dtype
        super().__init__(*args, **kwargs)

    def compose(self):
        fragment1 = Fragment1(
            self,
            name="fragment1",
            gpu_tensor=self.gpu_tensor,
            count=self.count,
            shape=self.shape,
            dtype=self.dtype,
        )
        fragment2 = Fragment2(self, name="fragment2")

        # Connect the two fragments (tx.out -> rx.in)
        # We can skip the "out" and "in" suffixes, as they are the default
        self.add_flow(fragment1, fragment2, {("tx", "rx")})


def main(on_gpu=False, count=10, shape=(64, 32), dtype=np.uint8, data_flow_tracking_enabled=False):
    app = MyPingApp(gpu_tensor=on_gpu, count=count, shape=shape, dtype=dtype)

    if data_flow_tracking_enabled:
        with Tracker(
            app,
            num_start_messages_to_skip=0,
            num_last_messages_to_discard=0,
        ) as trackers:
            # set separate log files for each fragment
            for fragment_name, tracker in trackers.items():
                tracker.enable_logging(fragment_name + "_logger.log")
            app.run()
            print(f"{type(trackers)=}, {trackers=}")
            for fragment_name, tracker in trackers.items():
                print(f"Fragment: {fragment_name}")
                tracker.print()
    else:
        app.run()


if __name__ == "__main__":
    description = (
        "Distributed tensor-based ping application. The tensor will have shape ([batch], rows, "
        "[columns], [channels]) where dimensions listed in square brackets are only present when "
        "the corresponding parameter is not 0."
    )
    parser = ArgumentParser(description=description, add_help=False)
    # Application().argv will intercept, -h and --help. Can provide --info instead
    # to see the help strings for these additional custom options.
    parser.add_argument(
        "--gpu",
        action="store_true",
        help="Use a GPU tensor instead of a host tensor",
    )
    parser.add_argument(
        "--count",
        type=int,
        default=10,
        help="The number of times to transmit the tensor.",
    )
    parser.add_argument(
        "--batch",
        type=int,
        default=0,
        help="The batch size of the tensor. If 0, no batch dimension will be present.",
    )
    parser.add_argument(
        "--rows",
        type=int,
        default=32,
        help="The number of rows for the transmitted tensor.",
    )
    parser.add_argument(
        "--columns",
        type=int,
        default=64,
        help=(
            "The number of columns for the transmitted tensor. If 0, no columns dimension will be "
            "present."
        ),
    )
    parser.add_argument(
        "--channels",
        type=int,
        default=0,
        help=(
            "The number of channels for the transmitted tensor. If 0, no channels dimension will "
            "be present."
        ),
    )
    parser.add_argument(
        "--data_type",
        type=str,
        default="uint8_t",
        help=(
            "The C++ data type of the tensor elements. It must be one of "
            "{'int8_t', 'int16_t', 'int32_t', 'int64_t', 'uint8_t', 'uint16_t', 'uint32_t', "
            " 'uint64_t', 'float', 'double', 'complex<float>', 'complex<double>'}"
        ),
    )
    parser.add_argument(
        "--track",
        action="store_true",
        help="Enable data flow tracking for the distributed app",
    )
    # use parse_known_args to ignore other CLI arguments that may be used by Application
    args, remaining = parser.parse_known_args()

    # can print the parser's help here prior to the Application's help output
    if "-h" in remaining or "--help" in remaining:
        print("\nAdditional arguments supported by this application:")
        print(textwrap.indent(parser.format_help(), "  "))
    else:
        # validate the arguments
        max_int32 = 2**31 - 1
        max_int64 = 2**63 - 1
        if args.count < 0 or args.count > max_int64:
            raise ValueError(f"count must be in range [0, {max_int64}]")
        if args.batch < 0 or args.batch > max_int32:
            raise ValueError(f"batch must be in range [0, {max_int32}]")
        if args.rows < 1 or args.rows > max_int32:
            raise ValueError(f"rows must be in range [1, {max_int32}]")
        if args.columns < 0 or args.columns > max_int32:
            raise ValueError(f"columns must be in range [0, {max_int32}]")
        if args.channels < 0 or args.channels > max_int32:
            raise ValueError(f"channels must be in range [0, {max_int32}]")

        tensor_str = "GPU" if args.gpu else "host"
        print(f"Configuring application to use {tensor_str} tensors")

    # only non-zero dimensions are included in the shape
    shape = tuple(s for s in (args.batch, args.rows, args.columns, args.channels) if s > 0)

    main(
        on_gpu=args.gpu,
        count=args.count,
        shape=shape,
        dtype=args.data_type,
        data_flow_tracking_enabled=args.track,
    )
